pykafka/producer.py

```diff
@@ -102,7 +124,32 @@ def __init__(self,
         self._retry_backoff_ms = retry_backoff_ms
         self._required_acks = required_acks
         self._ack_timeout_ms = ack_timeout_ms
-        self._batch_size = batch_size
+        self._max_queued_messages = max_queued_messages
+        self._min_queued_messages = min_queued_messages
+        self._linger_ms = linger_ms
+        self._block_on_queue_full = block_on_queue_full
+        self._worker_exception = None
+        self._worker_trace_logged = False
+        self._running = False
+        self.start()
+
+    def raise_worker_exceptions(fn):
```

Comment:
So, I'm actually not a fan of a decorator here.  This is a check which needs to happen, but it also looks like something which was sprinkled around without a ton of consideration as to when we're raising exceptions and how. It's called in places where it's detrimental (`stop`) and called places where it'll be ineffective (`_prepare_request` and `_send_request`). For the former case, it should be called after `self._running = False`, but you can't do that because it's a decorator.  In that latter case, it'll crash a background thread by raising an exception which won't be raised to the user and might kill an otherwise healthy thread.

This is why I don't like to decorator approach. It encourages adding it all over the place "just to be safe" because decorators look safe. Maybe it's just how I perceive it, but if this were a function call at the top of user-facing functions (i.e. a Guard), it'd be less likely to be added as it has been.

Maybe that's just me, though. Keep the decorator if you think it's useful, but right now it's used in way too many places. It should be used to guard user-facing functions, and in cases where it's not going to hurt things which otherwise should be happening (like shutting down the producer).